### PR TITLE
Display the coming soon page on front-end requests

### LINKED ISSUE #46222: [Launch Your Store] Coming Soon Routing
## Issue Description
Route to the coming soon page on front-end requests when applicable.

## Acceptance Criteria
* Logged-in users should not see the coming soon page. They see a banner handled in https://github.com/woocommerce/woocommerce/issues/46073
* When the site is in live mode, no coming soon page should display
* When the whole site is in coming soon mode, all pages should route to the coming soon page
* When only the store is in coming soon mode, only WooCommerce store pages should be protected by a coming soon page
* The coming soon page defined in advanced settings should be shown

### PR DESCRIPTION
**Pull request by adrianduffell**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Displays the coming soon page on applicable front-end requests when the site is in coming soon mode. Depending on the coming soon settings, either the whole site or just the store pages will be protected. It displays the coming soon page defined in the advanced settings.

Logged in admins do not see the coming soon page -- instead they will see a footer banner handled outside this PR.

Uses the [container](https://github.com/woocommerce/woocommerce/blob/c2f486b407883e6e7462ac5f8402647375c927d3/plugins/woocommerce/src/README.md#the-container) and dependency injection to load the classes.

Closes #46222  .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the WooCommerce Admin feature flag `launch-your-store`
2. Go to WooCommerce settings and set the site to coming soon mode.
3. As an admin user, see that the front-end is shown normally (or with a footer banner).
4. Visit the site home screen from a new private window.
5. See that the coming soon page is displayed.
6. In the WooCommerce settings, restrict coming soon to store pages only.
7. Reload the homescreen in the private window and see that the coming soon page is **not** displayed.
8. Navigate to a store page like the cart.
9. See that the coming soon page is displayed

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by adrianduffell** [comment]
There's a failing test in CI that passes for me locally 🤔 

```
1) Automattic\WooCommerce\Tests\ComingSoon\ComingSoonRequestHandlerTest::test_shop_manager_exclusion
Failed asserting that null is identical to 99.
```

It seems the admin exclusion does not work and they see the coming soon page. @woocommerce/ghidorah, is it worth skipping the test for now?

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		if ( $this->is_site_live() ) {
+			return false;
+		}
+
+		if ( $this->is_site_coming_soon() ) {
+			return true;
+		}
+
+		// Check the URL is a store page when in "store coming soon" mode.
+		if ( $this->is_store_coming_soon() && WCAdminHelper::is_store_page( $url ) ) {
```
Comment (rjchow) [inline plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php:48]: I had a look at the code for `is_store_page()` and I think we might want to do some error handling around this

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	 * Builds the relative URL from the WP instance.
+	 *
+	 * @internal
+	 * @link https://wordpress.stackexchange.com/a/274572
+	 * @param \WP $wp WordPress environment instance.
+	 */
+	public function get_url_from_wp( \WP $wp ) {
+		// Special case for plain permalinks.
+		if ( empty( get_option( 'permalink_structure' ) ) ) {
+			return '/' . add_query_arg( $wp->query_vars, $wp->request );
```
Comment (rjchow) [inline plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php:66]: This is pretty foreign to me but should we do some sanitisation here?

**Review by rjchow — APPROVED** [review]
Tests well as expected! I'm not familiar enough with the range of URLs that might exist, but testing the links available in the default store seems to work well enough. 

I left two comments for a bit more extra thoroughness, since this code runs on every request. I wonder if we should also wrap it all in a error catch and fail appropriately? I'm not sure if we should fail open or fail closed, though.

**Comment by rjchow** [comment]
> It seems the admin exclusion does not work and they see the coming soon page. @woocommerce/ghidorah, is it worth skipping the test for now?

I just tested with shop_manager roles and customer roles, they both seem to work as expected. I wonder why the tests don't work, but I'm ok with skipping the test for now

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * Handles the parse_request hook to determine whether the current page needs
+ * to be replaced with a comiing soon screen.
+ */
+class ComingSoonRequestHandler {
+
+	/**
+	 * NHelper dependency.
+	 *
+	 * @var string
```
Comment (chihsuan) [inline plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php]: Should be `ComingSoonHelper`?

```suggestion
	 * Coming soon helper.
	 *
	 * @var ComingSoonHelper
```

**Review by chihsuan — APPROVED** [review]
Thanks, tested well! 👍 

Just left a minor suggestion for doc comment.

> I just tested with shop_manager roles and customer roles, they both seem to work as expected. I wonder why the tests don't work, but I'm ok with skipping the test for now

+1, I agree.

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php (hunk 2 above)
Comment (adrianduffell) [reply plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php:66]: I think you might be right. I was assuming it was already sanitized, but [this](https://github.com/WordPress/wordpress-develop/blob/5afdac7e057bf4faf097079c0704f6e834fdce97/src/wp-includes/class-wp.php#L330-L333) indicates raw values are being copied over. If that's the case, there would be a risk of an attacker crafting a URL param to trick the site into showing live content 🤔  I'll investigate more in a follow-up issue.

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php (hunk 1 above)
Comment (adrianduffell) [reply plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php:48]: @rjchow what type of error handling do you have in mind?

**Comment by adrianduffell** [comment]
I skipped the flaky test in 7bffb76 and will address the feedback in follow-up issues to merge this sooner.

**Pull request merged**